### ISSUE TITLE
Allow model building outside of OnModelCreating

### ISSUE THREAD

**Issue by jagalves** [aspnet/EntityFrameworkCore#11738]
With EFCore 2, in order to have all mappings explicit, I create the model with an empty ConventionSet. 

This scenario causes an exception with EFCore 2.1 (both preview1 and preview2)

Don't know if this is an actual bug or just something that I'm required to add as a convention in order to have a 100% by code mapping. Maybe there's some documentation that I can't find?

```
System.ArgumentNullException : Value cannot be null.
Parameter name: relationalTypeMapping
Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
Microsoft.EntityFrameworkCore.Storage.Internal.TypeMappedRelationalParameter..ctor(String invariantName, String name, RelationalTypeMapping relationalTypeMapping, Nullable`1 nullable)
Microsoft.EntityFrameworkCore.Storage.Internal.RelationalParameterBuilder.AddParameter(String invariantName, String name, IProperty property)
Microsoft.EntityFrameworkCore.Storage.RelationalCommandBuilderExtensions.AddParameter(IRelationalCommandBuilder commandBuilder, String invariantName, String name, IProperty property)
Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.CreateStoreCommand()
Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(DbContext _, ValueTuple`2 parameters)
Microsoft.EntityFrameworkCore.Storage.Internal.NoopExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
TestProject.UnitTest1.Test1() in UnitTest1.cs
```

### Steps to reproduce

Test code attached. Switch between EFCore 2 and 2.1 preview in order to see the different behavior.

[TestProject.zip](https://github.com/aspnet/EntityFrameworkCore/files/1927883/TestProject.zip)

### Further technical details
EF Core version: 2.1 preview 2
Database Provider: Sqlite
Operating system: window 10 1803 17134.1 (and previous)
IDE: at the moment 15.7 preview 4

**Comment by ajcvickers** [comment]
@jagalves When using an empty convention set the model must still be constructed in a valid way--that is, with all required configuration including everything that is normally done by convention. In this case, there is a convention that creates type mappings, so if you don't run it, then you'll need to manually configure the type mapping for every property.

In general, it's hard to be sure everything is configured correctly and that the model is valid if you don't let any conventions run. We don't have any documentation on this because it is extremely advanced and not something we recommend people to do.

**Comment by jagalves** [comment]
Well, this scenario worked fine in 2.0... I have a full complex model over which I have full control on how the mapping is defined and how my domain classes and my database tables are created/related. I'm trying to avoid cases where I generate migrations and then I have to go back and add overrides  (like in my example userBuilder->Ignore(x=>x.NotMapped)) to avoid having columns and tables created by accident. This scenario is even worse if you use some other tool to do the migrations, by the way.

In other words, I'm trying to use what EF has to offer me, not override what it's trying to guess on my behalf. Is it possible to do that with the example code I provided?

**Comment by ajcvickers** [comment]
@jagalves By convention, EF asks the provider for the most appropriate type mapping and annotates the property with that type mapping. If you're not happy with this convention, then you will need to create the type mappings explicitly and annotate them on the model. If you are okay with the type mapper "guessing" the type mapping, then you could ask the type mapper explicitly for each mapping and annotate the model from that...but then you would be doing exactly what the convention does. So I suppose this all depends on the definition of "guess".

**Comment by ajcvickers** [comment]
@jagalves The issue is still open because it hasn't been triaged yet.

There are no conventions that will add types marked with NotMapped to the model. If you are seeing this, then please file an issue with full details so that we can investigate.

Column types are not the same as type mappings. Here is the code that in the convention:
```C#
foreach (var property in modelBuilder.Metadata.GetEntityTypes().SelectMany(e => e.GetDeclaredProperties()))
{
    property.Builder.HasAnnotation(
        CoreAnnotationNames.TypeMapping,
        _typeMappingSource.FindMapping(property),
        ConfigurationSource.Convention);
}
```
This is what you will need to do manually if not using the convention.

You can do things with conventions to create a set with just the ones you want, but this requires using internal code since conventions are not public yet, and still requires an advanced understanding of the EF Core model to ensure that you are building a valid model without using certain conventions.

**Comment by jagalves** [comment]
@ajcvickers 

> There are no conventions that will add types marked with NotMapped to the model. If you are seeing this, then please file an issue with full details so that we can investigate.

I called it that to be obvious to whoever reads the code that I don't want it to be mapped if I didn't do it explicitly. If you change the code to keep the default conventions (override OnModelCreating and call GetModel from there with the provided modelbuilder), the test will fail in the assert that checks that it's null. Same thing for properties with a class type: they will throw because I didn't set the primary key. And then I have to go to the mapping add them to a list of the ones I'm ignoring. I had a considerable list before I found out clearing ConventionSets was possible.

I'll try tomorrow your suggestion. Too late to think in my timezone. Thanks.

**Comment by ajcvickers** [comment]
Notes from triage:
* General consensus is that building a model and getting it "right" without any conventions is currently extremely difficult and not likely to be a pit of success
* Putting this on the backlog to consider what to do about this once conventions are made public (tracked by #214)
* There are at least two kinds of conventions--those that detect and impact the structure of the model, and those that ensure other parts of the model are then configured appropriately--for example, type mappings. At a high level, removing some of the first kind is not unreasonable, while removing some of the second kind is likely to cause issues. Therefore, we might want to:
  * Explicitly enable removing/configuring of only certain conventions
  * Provide an "empty" convention set that still runs the second kind of convention without the first
  * Look into mechanisms to finalize/validate a model after it has been built manually in some way.

**Comment by sadjadbp** [comment]
I'm hitting this error with something like explained in #4410, so we need a GetCompiled method that uses 

```
var conventionSet = SqlServerConventionSetBuilder.Build();
var modelBuilder = new ModelBuilder(conventionSet);
```

And now this project works with 2.0, but gives me this error when I upgrade to 2.1

**Comment by luigi68** [comment]
Hi, I have exactly the same issue posted by @sadjadbp when calling the DbContext.SaveChanges method.
I noted that doing something like 
var listOfChangedEntities = DbContext.ChangeTracker.Entries().Where(x => x.State == EntityState,Modified || x => x.State == EntityState.Added) the SaveChanges call works fine.
Any suggestion?

**Comment by merrycoder** [comment]
Same for me as for @sadjadbp - blocks me from upgrading to 2.1.

**Comment by almazsr** [comment]
We met the same problem, when tried to create `IModel` through `ModelBuilder`. By default EF Core caches `IModel`, thus all `DbContext`'s use one instance of it. If you need a new specific instance of `IModel` per each `DbContext` creation, create custom `IModelSource` to prevent caching of `IModel`:
```
    public class CreatePerCallModelSource : ModelSource
    {
        public CreatePerCallModelSource(ModelSourceDependencies dependencies) : base(dependencies)
        {
        }

        public override IModel GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
        {
            return CreateModel(context, conventionSetBuilder, validator);
        }
    }
```
Then just replace the default ModelSource with created one to create a new `DbContext` with `DbContextOptionsBuilder`:
```dbContextOptionsBuilder.ReplaceService<IModelSource, CreatePerCallModelSource>();```

**Comment by luigi68** [comment]
> We met the same problem, when tried to create `IModel` through `ModelBuilder`. By default EF Core caches `IModel`, thus all `DbContext`'s use one instance of it. If you need a new specific instance of `IModel` per each `DbContext` creation, create custom `IModelSource` to prevent caching of `IModel`:
> 
> ```
>     public class CreatePerCallModelSource : ModelSource
>     {
>         public CreatePerCallModelSource(ModelSourceDependencies dependencies) : base(dependencies)
>         {
>         }
> 
>         public override IModel GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder, IModelValidator validator)
>         {
>             return CreateModel(context, conventionSetBuilder, validator);
>         }
>     }
> ```
> Then just replace the default ModelSource with created one to create a new `DbContext` with `DbContextOptionsBuilder`:
> `dbContextOptionsBuilder.ReplaceService<IModelSource, CreatePerCallModelSource>();`

I tried without luck...  honestly i believe problem is somewhere else.

**Comment by sabo3000** [comment]
same issue. please fix it as soon as possible!

**Comment by alekrist** [comment]
Having the same issue and this blocks us from upgrading to v2.1. Please fix this soon.

**Comment by Joganoth** [comment]
same issue for me, please fix it asap!

**Comment by ajcvickers** [comment]
@sabo3000 @alekrist @Joganoth  @mfernandezruiz @AAltemus @luigi68 @almazsr This issue is about the behavior of building the model using an explicitly empty convention set. Can you conform either:
* Confirm that this is the case for the issues you are seeing and provide some details as to why you are using an empty convention set
* Or file a new issue with a runnable project/solution that demonstrates the behavior you are seeing when using the normal conventions.

**Comment by mfernandezruiz** [comment]
This code runs until version 2.1 and with the 2.1.3 fails


using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Conventions;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace EFCoreTest
{
    class Program
    {
        static void Main(string[] args)
        {
            var conventions = SqliteConventionSetBuilder.Build();
            var modelBuilder = new ModelBuilder(conventions);
            modelBuilder.Entity<User>();
            var model = modelBuilder.Model;

            var optionsBuilder = new DbContextOptionsBuilder();
            optionsBuilder.UseSqlite("Data Source=Test.db");
            optionsBuilder.UseModel(model);

            using (var context = new MyContext(optionsBuilder.Options))
            {
                context.Database.EnsureCreated();

                var user = new User()
                {
                    Name = "Adam"
                };

                context.Add(user);
                context.SaveChanges();
            }
        }
    }

    public class MyContext : DbContext
    {
        public MyContext(DbContextOptions options) : base(options)
        {
        }
    }

    [Table("Users")]
    public class User
    {
        [Key]
        public int Id { get; set; }
        public string Name { get; set; }
    }
}

**Comment by luigi68** [comment]
> @sabo3000 @alekrist @Joganoth @mfernandezruiz @AAltemus @luigi68 @almazsr This issue is about the behavior of building the model using an explicitly empty convention set. Can you either:
> 
> * Confirm that this is the case for the issues you are seeing and provide some details as to why you are using an empty convention set
> * Or file a new issue with a runnable project/solution that demonstrates the behavior you are seeing when using the normal conventions.

About me I think I am not using empty convention set.
The following code virtually represent my actual implementation:
var connection = System.Data.SqlClient.SqlClientFactory.Instance.CreateConnection(...);
var conventionSet = SqlServerConventionSetBuilder.Build();
var modelBuilder = new ModelBuilder(conventionSet);

modelBuilder.Entity("ADDRESS").HasKey( )... // example...
...

var optionsBuilder = new DbContextOptionsBuilder();
optionsBuilder.UseModel(modelBuilder.Model);
optionsBuilder.UseSqlServer(connection);
...
var dbContext = new DbContext(optionsBuilder.Options)
...
Please let me know what i am doing wrong..
Thanks in advance

**Comment by luigi68** [comment]
Because my model is built one time and used many times. WebAPI rest: on the first call build the model and  use it on any subsequent calls (improving performance)

**Comment by neris** [comment]
@luigi68 Since EF Core 2.0 there is a feature called [DbContext pooling](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-2.0#dbcontext-pooling). If you configure your EF Core inside your application correctly, it will pool DbContexts like in a similar way you have your connections to the database pooled. It creates a DbContext once (or a couple of them) and then reuses across your application.

**Comment by luigi68** [comment]
I explain briefly: I created a framework that can use  EF or EF Core (I was not able to adopt only EF Core because of missing database drivers - ie. Oracle, SAP...). 
The framework works correctly for EF/EFCore aside when I SaveChanges()... 
In my application connection is created externally to the Entity Framework passed to it and closed and the end of the WebCall and I don't want DbContext in the pooling..

**Comment by AAltemus** [comment]
I was doing something that was using  `var modelBuilder = new ModelBuilder(SqlServerConventionSetBuilder.Build());` and using `modelBuilder.Entity(entityType)` where entityType was a DbContext type, then trying to return using `modelBuild.Model` .   I was only using annotations at this point.  I could not use the OnModelCreating because this is a Singleton that was going to be re-used.  

Changing it to use an EntityTypeConfiguration and applying it, _then_ calling `model.Validate()` and `sqlServerModelValidator.Validate(model)`  then returning model fixed it for for me. 

Apologies if this was the wrong place to put this.  I was racking my brain trying to find any information regarding this issue and this was the only thing I could find on it.

**Comment by merrycoder** [comment]
We are more or less using the same code as @mfernandezruiz below. No matter if we use empty convention, or conventions built with SqliteConventionSetBuilder or SqlServerConventionSetBuilder, the code fails with the indicated error. We are not using database migrations, in case this makes a difference.

> This code runs until version 2.1 and with the 2.1.3 fails
> 
> ```
> using Microsoft.EntityFrameworkCore;
> using Microsoft.EntityFrameworkCore.Metadata.Conventions;
> using System.ComponentModel.DataAnnotations;
> using System.ComponentModel.DataAnnotations.Schema;
> 
> namespace EFCoreTest
> {
>     class Program
>     {
>         static void Main(string[] args)
>         {
>             var conventions = SqliteConventionSetBuilder.Build();
>             var modelBuilder = new ModelBuilder(conventions);
>             modelBuilder.Entity<User>();
>             var model = modelBuilder.Model;
> 
>             var optionsBuilder = new DbContextOptionsBuilder();
>             optionsBuilder.UseSqlite("Data Source=Test.db");
>             optionsBuilder.UseModel(model);
> 
>             using (var context = new MyContext(optionsBuilder.Options))
>             {
>                 context.Database.EnsureCreated();
> 
>                 var user = new User()
>                 {
>                     Name = "Adam"
>                 };
> 
>                 context.Add(user);
>                 context.SaveChanges();
>             }
>         }
>     }
> 
>     public class MyContext : DbContext
>     {
>         public MyContext(DbContextOptions options) : base(options)
>         {
>         }
>     }
> 
>     [Table("Users")]
>     public class User
>     {
>         [Key]
>         public int Id { get; set; }
>         public string Name { get; set; }
>     }
> }
> ```

**Comment by mfernandezruiz** [comment]
> @mfernandezruiz, @luigi68 You can use a standard way of creating a model by overriding the `OnModelCreating` method in your `DbContext`: https://docs.microsoft.com/en-us/ef/core/modeling/
> Looks like you don't need the actual functionality this issue addresses. I don't see any convention changes in your examples.


My example is just an example. My application needs to create the data models at run time and also needs to handle several data models at the same time.

**Comment by mfernandezruiz** [comment]
@neris : My example is only for reproduce the problem. My application needs to create data models at run time and also needs to handle several data models at the same time, so overwriting 'OnModelCreating' does not help

**Comment by ajcvickers** [comment]
Triage: we will look into adding a `Finalize`/`Build` or similar method to the model that will make it possible to build the model externally from the context but still have all the conventions execute.

**Comment by AndriySvyryd** [comment]
@almazsr If you need to rebuild the model based on some external information you can use implement `IModelCacheKeyFactory` so that it's only cached when appropriate, see [Alternating models with the same DbContext](https://docs.microsoft.com/en-us/ef/core/modeling/dynamic-model)

**Comment by smitpatel** [comment]
Also broke compat tests.

**Comment by stefan-landgraf-is24** [comment]
@ajcvickers Thanks for supporting our case. Maybe i misunderstood something, but I still get the same error with EFCore 2.2 and the new method FinalizeModel():

```
var mb = new ModelBuilder(new ConventionSet(){});
 foreach (var assembly in modelAssemblies)
 {
      mb.AddConfigurationsFromAssembly(assembly);
 }
mb.FinalizeModel();

var optionsBuilder = new DbContextOptionsBuilder<EFDataContext>()
   .UseModel(mb.Model)
```

leads to this exception when saving changes:

```
System.ArgumentNullException : Value cannot be null.
Parameter name: relationalTypeMapping
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
   at Microsoft.EntityFrameworkCore.Storage.Internal.TypeMappedRelationalParameter..ctor(String invariantName, String name, RelationalTypeMapping relationalTypeMapping, Nullable`1 nullable)
```

**Comment by stefan-landgraf-is24** [comment]
@ajcvickers  How can I find out what has to be done, if the error method gives no hint at all?

`System.ArgumentNullException : Value cannot be null.
Parameter name: relationalTypeMapping`

I have tried with an empty convetion set and with SqlServerConventionSet, both generate the same crypric errormessage.

**Comment by AndriySvyryd** [comment]
@stefan-landgraf-is24 You can use `SqlServerConventionSetBuilder.Build()`
